False positive: Statement is [unreachable] in loop with iteration-carried flag

**Bug Report**

mypy reports `Statement is unreachable` in a loop that uses a “skip next line” flag across iterations. The flagged branch is reachable in the next iteration, but mypy appears to analyze it as if the flag can’t be True.

**To Reproduce**

```python
from __future__ import annotations

import re

_HDR_BY_CMD = re.compile(r"^\s*#\s*by the following command:\s*$", re.I)
_HDR_CMD    = re.compile(r"^\s*#\s*pip-compile\b", re.I)
_ANY_VIA    = re.compile(r"^\s*#\s*via\b.*$", re.I)

def norm(lines: list[str]) -> list[str]:
    out: list[str] = []
    skip_next = False
    for raw in lines:
        ln = raw.rstrip()
        if skip_next and _HDR_CMD.match(ln):
            skip_next = False
            continue
        if _HDR_BY_CMD.match(ln):
            skip_next = True
            continue
        if _ANY_VIA.match(ln):
            out.append("# via")
        else:
            out.append(ln)
    return out
```

Command:
mypy --show-error-codes --warn-unreachable demo.py

Expected Behavior
No unreachable error; the first if is reachable on iterations after _HDR_BY_CMD matched.

Actual Behavior
demo.py:20: error: Statement is unreachable  [unreachable]
Found 1 error in 1 file (checked 1 source file)

Environment

mypy 1.18.1 (compiled: yes)

Python version used: 3.13.7

OS: Windows 11 (Git Bash / PowerShell 7.5.3)

Command-line flags: --warn-unreachable --show-error-codes

Config
[tool.mypy]
python_version = "3.13"
warn_unreachable = true

Project repo error encountered in
https://github.com/sMiNT0S/AIBugBench/tree/main ;
https://github.com/sMiNT0S/AIBugBench/pull/18
(specifically, 'update_requirements_lock.py' lines 200 ~224)

--- a/mypy/typeops.py
+++ b/mypy/typeops.py
@@ -791,7 +791,7 @@ def false_only(t: Type) -> ProperType:
         elif isinstance(t, Instance):
             if (t.type.is_final or t.type.is_enum) and state.strict_optional:
                 return UninhabitedType(line=t.line)
-        elif isinstance(t, LiteralType) and t.is_enum_literal():
+        elif isinstance(t, LiteralType) and t.is_enum_literal() and state.strict_optional:
             return UninhabitedType(line=t.line)
 
         new_t = copy_type(t)

--- a/test-data/unit/check-inference.test
+++ b/test-data/unit/check-inference.test
@@ -1976,6 +1976,19 @@ if not x:
     x = C()
 [builtins fixtures/dict.pyi]
 
+[case testNoWrongUnreachableWarningWithNoStrictOptionalAndEnumLiteral]
+# flags: --no-strict-optional --warn-unreachable
+from enum import Enum
+from typing import Literal, Optional
+
+class E(Enum):
+    a = 1
+
+x: Optional[Literal[E.a]]
+if not x:
+    x = E.a
+[builtins fixtures/dict.pyi]
+
 [case testInferFromEmptyListWhenUsingInWithStrictEquality]
 # flags: --strict-equality
 def f() -> None:

